pkg/status/status.go

```diff
@@ -0,0 +1,536 @@
+	url          *url.URL
+	podIP        string
+	podPort      string
+	logger       *zap.SugaredLogger
+}
+
+// ProbeTarget contains the URLs to probes for a set of Pod IPs serving out of the same port.
+type ProbeTarget struct {
+	PodIPs  sets.Set[string]
+	PodPort string
+	Port    string
+	URLs    []*url.URL
+}
+
+type ProbeState struct {
+	Version string
+	Ready   bool
+}
+
+type Backends struct {
+	Key        types.NamespacedName
+	Version    string
+	URLs       map[Visibility]URLSet
+	HTTPOption v1alpha1.HTTPOption
+}
+
+func (b *Backends) AddURL(v Visibility, u url.URL) {
+	if b.URLs == nil {
+		b.URLs = make(map[Visibility]URLSet)
+	}
+	urls, ok := b.URLs[v]
+	if !ok {
+		urls = make(URLSet)
+		b.URLs[v] = urls
+	}
+	urls.Insert(u)
+}
+
+type Visibility = v1alpha1.IngressVisibility
+type URLSet = sets.Set[url.URL]
+
+// ProbeTargetLister lists all the targets that requires probing.
+type ProbeTargetLister interface {
+	// BackendsToProbeTargets produces list of targets for the given backends
+	BackendsToProbeTargets(ctx context.Context, backends Backends) ([]ProbeTarget, error)
+}
+
+// Manager provides a way to check if an Ingress is ready
+type Manager interface {
+	DoProbes(ctx context.Context, backends Backends) (ProbeState, error)
+	IsProbeActive(ing *v1alpha1.Ingress) (ProbeState, bool)
+}
+
+// Prober provides a way to check if a VirtualService is ready by probing the Envoy pods
+// handling that VirtualService.
+type Prober struct {
+	logger *zap.SugaredLogger
+
+	// mu guards ingressStates and podContexts
+	mu            sync.Mutex
+	ingressStates map[types.NamespacedName]*ingressState
+	podContexts   map[string]cancelContext
+
+	workQueue workqueue.RateLimitingInterface
+
+	targetLister ProbeTargetLister
+
+	readyCallback func(types.NamespacedName)
+
+	probeConcurrency int
+}
+
+// NewProber creates a new instance of Prober
+func NewProber(
+	logger *zap.SugaredLogger,
+	targetLister ProbeTargetLister,
+	readyCallback func(types.NamespacedName)) *Prober {
+	return &Prober{
+		logger:        logger,
+		ingressStates: make(map[types.NamespacedName]*ingressState),
+		podContexts:   make(map[string]cancelContext),
+		workQueue: workqueue.NewNamedRateLimitingQueue(
+			workqueue.NewMaxOfRateLimiter(
+				// Per item exponential backoff
+				workqueue.NewItemExponentialFailureRateLimiter(50*time.Millisecond, 30*time.Second),
+				// Global rate limiter
+				&workqueue.BucketRateLimiter{Limiter: rate.NewLimiter(rate.Limit(50), 100)},
+			),
+			"ProbingQueue"),
+		targetLister:     targetLister,
+		readyCallback:    readyCallback,
+		probeConcurrency: probeConcurrency,
+	}
+}
+
+func (m *Prober) IsProbeActive(key types.NamespacedName) (ProbeState, bool) {
+	state, ok := func() (ProbeState, bool) {
+		m.mu.Lock()
+		defer m.mu.Unlock()
+		if state, ok := m.ingressStates[key]; ok {
+			return ProbeState{Version: state.version, Ready: state.pendingCount.Load() == 0}, true
+		}
+		return ProbeState{}, false
+	}()
+
+	return state, ok
+}
+
```

Comment:
Do you mind adding a description to the DoProbes function like IsReady had before?

Reply:
done